technology_and_tooling/testing/testable_code_fixtures.md

```diff
@@ -0,0 +1,799 @@
+
+
+::::challenge{id=test-patient-class title="Grouping tests in a class."}
+
+ Encapsulate the tests for the `Patient` class inside a class named `TestPatient`. Include a method `setup_class` where the two `Patient` objects (with `id` of 1 and 2) will be created rather than creating an object within each test.
+
+:::solution
+~~~python
+import pytest
+from inflammation.models import Patient
+
+class TestPatient:
+    def setup_class(self):
+        self.patient1 = Patient(id=1, data=[1, 2, 3, 4, 5])
+        self.patient2 = Patient(id=2, data=[10, 20, 30, 40, 50])
+
+    def test_patient_data_mean(self):
+        assert self.patient1.data_mean() == 3.0
+
+    def test_patient_data_max(self):
+        assert self.patient1.data_max() == 5
+
+    def test_patient_data_min(self):
+        assert self.patient1.data_min() == 1
+
+    def test_patient_attributes(self):
+        assert self.patient2.id == 2
+        assert self.patient2.data == [10, 20, 30, 40, 50]
+
+~~~
+
+:::
+::::
+
+:::callout{variant="note"}
+## Fixtures
+
+As an alternative to encapsulating test methods in a class and using `setup` and `teardown` methods, we can use *fixtures*. Fixtures are defined by using the `@pytest.fixture` decorator on a function. This function will then become available to be passed as an argument to your tests and used within them. If there is a cleanup part to the code, then the fixture function should use a `yield` statement rather than a `return` statement. Anything up to the `yield` statement is setup code, and anything after the statement will be run post-testing in order to clean up (teardown code).
+
+Here is how we can write our tests for the `Person` class using fixtures instead of a `setup_class` method:
+
+~~~python
+import pytest
+from inflammation.models import Patient
+
+@pytest.fixture()
+def patient_1():
+    return Patient(id=1, data=[1, 2, 3, 4, 5])
+
+@pytest.fixture()
+def patient_2():
+    return Patient(id=2, data=[10, 20, 30, 40, 50])
+
+def test_patient_data_mean(patient_1):
+    assert patient_1.data_mean() == 3.0
+
+def test_patient_data_max(patient_1):
+    assert patient_1.data_max() == 5
+
+def test_patient_data_min(patient_1):
+    assert patient_1.data_min() == 1
+
+def test_patient_attributes(patient_2):
+    assert patient_2.id == 2
+    assert patient_2.data == [10, 20, 30, 40, 50]
+~~~
+
+We could also encapsulate our tests in a class as well as using fixtures if we wished in order to keep them organised. In this case, the fixtures are available to be used within any classes we create in the file:
```

Comment:
I think, since tests are tests on the same class you would always write them within a test class. And since this callout is quite long, it might be better to remove the first example and just keep the second, what do you think/?

Reply:
Have removed the first example, thanks!